compaction/compaction_strategy.cc

```diff
@@ -305,15 +304,53 @@ class time_window_backlog_tracker final : public compaction_backlog_tracker::imp
             }
         }
 
+        ssize_t undo_count = 0;
+        auto undo = defer([&] () noexcept {
+            clogger.warn("time_window_backlog_tracker: replace_sstables failed. Rolling back...");
+            for (const auto& [bound, r] : per_window_replacement) {
+                if (undo_count-- <= 0) {
+                    break;
+                }
+                // All windows must exist here.
+                // 1. for new_ssts, the window is created before undo_count is incremented.
+                // 2. for old_ssts:
+                //    a. if the window did not exist when entring, they would not be registered in per_window_replacement.
+                //    b. if the window exists and no sstables remain in it after w.replace_sstables, it is only marked with r.do_erase
+                //       and the window is erased only after this deferred undo function is canceled.
+                auto it = _windows.find(bound);
+                if (it == _windows.end()) {
+                    on_fatal_internal_error(clogger, fmt::format("No window found when undoing replace_sstables for bound={} old_ssts={} new_ssts={}", bound, old_ssts, new_ssts));
+                }
+                auto& w = it->second;
+                w.replace_sstables(r.new_ssts, r.old_ssts);
```

Comment:
But, this can throw, no? And we mustn't throw from a destructor.

Reply:
Better to prepare the new state on the side and move it once all is done than attempt to undo and fall into the same trap.

Reply:
okay

Reply:
done in v6